crates/accelerate/src/circuit_library/entanglement.rs

```diff
@@ -0,0 +1,202 @@
+/// [(0..m-1), (1..m), (2..m+1), ..., (n-m..n-1)]
+pub fn linear(num_qubits: u32, block_size: u32) -> impl DoubleEndedIterator<Item = Vec<u32>> {
+    (0..num_qubits - block_size + 1)
+        .map(move |start_index| (start_index..start_index + block_size).collect())
+}
+
+/// Get a reversed linear entanglement. This is like linear entanglement but in reversed order:
+/// [(n-m..n-1), ..., (1..m), (0..m-1)]
+/// This is particularly interesting, as CX+"full" equals CX+"reverse_linear", even though the
+/// latter uses a quadratically less number of gates.
+pub fn reverse_linear(num_qubits: u32, block_size: u32) -> impl Iterator<Item = Vec<u32>> {
+    linear(num_qubits, block_size).rev()
+}
+
+/// Return the qubit indices for circular entanglement. This is defined as tuples of length ``m``
+/// starting at each possible index ``(0..n)``. Historically, Qiskit starts with index ``n-m+1``.
+/// This is probably easiest understood for a concerete example of 4 qubits and block size 3:
+/// [(2,3,0), (3,0,1), (0,1,2), (1,2,3)]
+pub fn circular(num_qubits: u32, block_size: u32) -> Box<dyn Iterator<Item = Vec<u32>>> {
+    if block_size == 1 || num_qubits == block_size {
+        Box::new(linear(num_qubits, block_size))
+    } else {
+        let historic_offset = num_qubits - block_size + 1;
+        Box::new((0..num_qubits).map(move |start_index| {
+            (0..block_size)
+                .map(|i| (historic_offset + start_index + i) % num_qubits)
+                .collect()
+        }))
+    }
+}
+
+/// Get pairwise entanglement. This is typically used on 2 qubits and only has a depth of 2, as
+/// first all odd pairs, and then even pairs are entangled. For example on 6 qubits:
+/// [(0, 1), (2, 3), (4, 5), /* now the even pairs */ (1, 2), (3, 4)]
+pub fn pairwise(num_qubits: u32) -> impl Iterator<Item = Vec<u32>> {
+    // for Python-folks (like me): pairwise is equal to linear[::2] + linear[1::2]
+    linear(num_qubits, 2)
+        .step_by(2)
+        .chain(linear(num_qubits, 2).skip(1).step_by(2))
+}
+
+/// The shifted, circular, alternating (sca) entanglement is motivated from circuits 14/15 of
+/// https://arxiv.org/abs/1905.10876. It corresponds to circular entanglement, with the difference
+/// that in each layer (controlled by ``offset``) the entanglement gates are shifted by one, plus
+/// in each second layer, the entanglement gate is turned upside down.
+/// Offset 0 -> [(2,3,0), (3,0,1), (0,1,2), (1,2,3)]
+/// Offset 1 -> [(3,2,1), (0,3,2), (1,0,3), (2,1,0)]
+/// Offset 2 -> [(0,1,2), (1,2,3), (2,3,0), (3,0,1)]
+/// ...
+pub fn shift_circular_alternating(
+    num_qubits: u32,
+    block_size: u32,
+    offset: usize,
+) -> Box<dyn Iterator<Item = Vec<u32>>> {
+    // index at which we split the circular iterator -- remember that circular entanglement is a
+    // list of length ``num_qubits``, which is what we use in the ``convert_idx`` function here
+    let split = PySequenceIndex::convert_idx(
```

Comment:
I think it's better to maintain a cleaner separation between the implementation of Qiskit data and functions in Rust on the one hand, and on the Python wrappers on the other. Separating them in the future will be a very large undertaking even if we try to prepare for it.

Reply:
The ``convert_idx`` function doesn't actually use any `pyo3` or Python logic, it's really just a util for Python-like access. I can also avoid using it but if we have the function already, then I'm not sure we should duplicate the logic 🙂

Reply:
> The `convert_idx function` doesn't actually use any `pyo3`

That's part of the point I'm trying to make. When I first started to the comment, I mistakenly believed that  `convert_idx` was introduced in this PR. This makes dealing with this a bit less straightforward. `convert_idx` uses no `pyo3`. The two calls to `convert_idx` are in functions that otherwise do not use pyo3. Part of what we are trying to do is redesign qiskit in a way that is independent of Python. Allowing pyo3 to intrude where it is not necessary makes this more difficult.

Reply:
Ok, this seems part of a more general discussion -- to move forward I removed the usage of `convert_idx` from this PR 🙂